libsql-sqlite3/src/vectorfloat32.c

```diff
@@ -0,0 +1,238 @@
+** Utility routines for vector serialization and deserialization
+**************************************************************************/
+
+static inline unsigned formatF32(float value, char *pBuf, int nBufSize){
+  sqlite3_snprintf(nBufSize, pBuf, "%g", (double)value);
+  return strlen(pBuf);
+}
+
+static inline unsigned serializeF32(unsigned char *pBuf, float value){
+  u32 *p = (u32 *)&value;
+  pBuf[0] = *p & 0xFF;
+  pBuf[1] = (*p >> 8) & 0xFF;
+  pBuf[2] = (*p >> 16) & 0xFF;
+  pBuf[3] = (*p >> 24) & 0xFF;
+  return sizeof(float);
+}
+
+static inline float deserializeF32(const unsigned char *pBuf){
+  u32 value = 0;
+  value |= (u32)pBuf[0];
+  value |= (u32)pBuf[1] << 8;
+  value |= (u32)pBuf[2] << 16;
+  value |= (u32)pBuf[3] << 24;
+  return *(float *)&value;
+}
+
+size_t vectorF32SerializeToBlob(
+  const Vector *pVector,
+  unsigned char *pBlob,
+  size_t nBlobSize
+){
+  float *elems = pVector->data;
+  unsigned char *pPtr = pBlob;
+  size_t len = 0;
+  unsigned i;
+
+  assert( pVector->type == VECTOR_TYPE_FLOAT32 );
+  assert( pVector->dims <= MAX_VECTOR_SZ );
+  assert( nBlobSize >= pVector->dims * sizeof(float) );
+
+  for(i = 0; i < pVector->dims; i++){
+    pPtr += serializeF32(pPtr, elems[i]);
+    len += sizeof(float);
+  }
+  return len;
+}
+
+size_t vectorF32DeserializeFromBlob(
+  Vector *pVector,
+  const unsigned char *pBlob,
+  size_t nBlobSize
+){
+  float *elems = pVector->data;
+  unsigned i;
+  pVector->type = VECTOR_TYPE_FLOAT32;
+  pVector->dims = nBlobSize / sizeof(float);
+
+  assert( pVector->dims <= MAX_VECTOR_SZ );
+  assert( nBlobSize >= vectorDataSize(pVector->type, pVector->dims) );
+
+  for(i = 0; i < pVector->dims; i++){
+    elems[i] = deserializeF32(pBlob);
+    pBlob += sizeof(float);
+  }
+  return vectorDataSize(pVector->type, pVector->dims);
+}
+
+void vectorF32Serialize(
+  sqlite3_context *context,
+  const Vector *pVector
+){
+  float *elems = pVector->data;
+  unsigned char *pBlob;
+  size_t nBlobSize;
+
+  assert( pVector->type == VECTOR_TYPE_FLOAT32 );
+  assert( pVector->dims <= MAX_VECTOR_SZ );
+
+  nBlobSize = vectorDataSize(pVector->type, pVector->dims);
+
+  if( nBlobSize == 0 ){
+    sqlite3_result_zeroblob(context, 0);
+    return;
+  }
+
+  pBlob = sqlite3_malloc64(nBlobSize);
+  if( !pBlob ){
+    sqlite3_result_error_nomem(context);
+    return;
+  }
+
+  vectorF32SerializeToBlob(pVector, pBlob, nBlobSize);
+  sqlite3_result_blob(context, (char*)pBlob, nBlobSize, sqlite3_free);
+}
+
+#define SINGLE_FLOAT_CHAR_LIMIT 32
+void vectorF32MarshalToText(
+  sqlite3_context *context,
+  const Vector *pVector
+){
+  float *elems = pVector->data;
+  size_t nBufSize;
+  size_t iBuf = 0;
+  char *pText;
+  char valueBuf[SINGLE_FLOAT_CHAR_LIMIT];
+
+  assert( pVector->type == VECTOR_TYPE_FLOAT32 );
+  assert( pVector->dims <= MAX_VECTOR_SZ );
+
+  nBufSize = 2 + pVector->dims * (SINGLE_FLOAT_CHAR_LIMIT + 1 /* plus comma */);
```

Comment:
Do we put comma after last element as well?

Reply:
It seems not.

Reply:
but I guess this extra byte is used for \0 - is that right?

Reply:
No, we don't put. I think here we allocate 1 more extra unnecessary byte (it is really unnecessary since `sqlite3_result_text` ask for blob length explicitly). But I guess it's fine.

Reply:
It is fine. Probably a comment would help to avoid confusion of the reader.
Also it seems that we can't remove that one extra char because it would make the lenght odd and this check would fail https://github.com/tursodatabase/libsql/pull/1531/files#diff-98fd1b40c774432f5288bd6ae7000632bbffb8687f15734977cb8cc794ed1a41R236